cdap-master/src/main/java/co/cask/cdap/data/tools/MDSUpgrader.java

```diff
@@ -0,0 +1,347 @@
+      public Void apply(DatasetContext<Table> input) throws Exception {
+        Scanner rows = input.get().scan(partialKey, Bytes.stopKeyForPrefix(partialKey));
+        Row row;
+        while ((row = rows.next()) != null) {
+          RunRecord runRecord = GSON.fromJson(Bytes.toString(row.get(COLUMN)), RunRecord.class);
+
+          MDSKey.Splitter keyParts = new MDSKey(row.getRow()).split();
+          // skip runRecordStarted
+          keyParts.getString();
+          // skip default
+          keyParts.skipString();
+          // skip appId
+          keyParts.skipString();
+          // skip programId
+          keyParts.skipString();
+          long startTs = keyParts.getLong();
+          String pId = keyParts.getString();
+
+          // we cannot add a runRecordCompleted if there is no runRecordStarted for the run so write it first
+          // the next setStop call deletes the runRecordStarted for that run so we don't have to delete it here.
+          writeTempRunRecordStart(appId, programType, programId, pId, startTs);
+          store.setStop(Id.Program.from(Id.Application.from(Constants.DEFAULT_NAMESPACE, appId), programType,
+                                        programId), pId, runRecord.getStopTs(),
+                        getControllerStateByStatus(runRecord.getStatus()));
+        }
+        return null;
+      }
+    });
+  }
+
+  /**
+   * Writes the {@link AppMetadataStore#TYPE_RUN_RECORD_STARTED} entry in the app meta table so that
+   * {@link AppMetadataStore#TYPE_RUN_RECORD_COMPLETED} can be written which deleted the started record.
+   *
+   * @param appId       the application id
+   * @param programType {@link ProgramType} of this program
+   * @param programId   the program id of this program
+   * @param pId         the process id of the run
+   * @param startTs     the startTs
+   */
+  private void writeTempRunRecordStart(String appId, ProgramType programType, String programId, String pId,
+                                       long startTs) {
+    store.setStart(Id.Program.from(Id.Application.from(Constants.DEFAULT_NAMESPACE, appId), programType, programId),
+                   pId, Long.MAX_VALUE - startTs);
+  }
+
+  /**
+   * Writes the updated application metadata through the {@link DefaultStore}
+   *
+   * @param row the {@link Row} containing the application metadata
+   * @throws URISyntaxException if failed to create {@link URI} from the archive location in metadata
+   */
+  private void appMetaUpgrader(Row row) throws URISyntaxException, IOException {
+    ApplicationMeta appMeta = GSON.fromJson(Bytes.toString(row.get(COLUMN)), ApplicationMeta.class);
+    store.addApplication(Id.Application.from(Constants.DEFAULT_NAMESPACE, appMeta.getId()), appMeta.getSpec(),
+                         updateAppArchiveLocation(appMeta.getId(), new URI(appMeta.getArchiveLocation())));
+    // store the stream names which belongs to this app as we want to filter out streams which do not belong to any app
```

Comment:
Ok, my main concern was that in the future, streams in an app in one namespace would stop all streams in other namespaces from being upgraded, since you're only storing String to identify the stream.
Should use Id.Stream, to also work in the future.